File: DSAL/F/F26.java
// 26. Assume we have two input and two output tapes to perform the sorting. The internal memory can hold and sort m records at a time. Write a program in java for external sorting. Find out time complexity.

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.InputStreamReader;

public class F26 {
	static final String INPUT_FILE = "F26_Input.txt";
	static final String T1 = "F26_T1.txt";
	static final String T2 = "F26_T2.txt";
	static final String T3 = "F26_T3.txt";
	static final String T4 = "F26_T4.txt";
	static final String OUTPUT_FILE = "F26_Sorted.txt";

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

	static int readInt(String prompt) throws Exception {
		while (true) {
			System.out.print(prompt);
			String s = br.readLine();
			if (s == null) return 0;
			s = s.trim();
			try {
				return Integer.parseInt(s);
			} catch (Exception e) {
				System.out.println("Invalid number. Try again.");
			}
		}
	}

	static Integer readNextInt(BufferedReader r) throws Exception {
		while (true) {
			String line = r.readLine();
			if (line == null) return null;
			line = line.trim();
			if (line.length() == 0) continue;
			try {
				return Integer.valueOf(Integer.parseInt(line));
			} catch (Exception e) {
				// skip malformed line
			}
		}
	}

	static int countRecords(String path) throws Exception {
		File f = new File(path);
		if (!f.exists()) return 0;
		BufferedReader r = new BufferedReader(new FileReader(f));
		int n = 0;
		while (readNextInt(r) != null) n++;
		r.close();
		return n;
	}

	static boolean fileHasAnyInt(String path) throws Exception {
		File f = new File(path);
		if (!f.exists()) return false;
		BufferedReader r = new BufferedReader(new FileReader(f));
		boolean ok = (readNextInt(r) != null);
		r.close();
		return ok;
	}

	static void displayFile(String path, String title) throws Exception {
		System.out.println("\n--- " + title + " ---");
		File f = new File(path);
		if (!f.exists()) {
			System.out.println("(file not found)");
			return;
		}
		BufferedReader r = new BufferedReader(new FileReader(f));
		int shown = 0;
		Integer x;
		while ((x = readNextInt(r)) != null) {
			System.out.print(x + " ");
			shown++;
			if (shown % 20 == 0) System.out.println();
		}
		r.close();
		if (shown == 0) System.out.print("(empty)");
		System.out.println("\nTotal records: " + shown);
	}

	static void selectionSort(int[] arr, int n) {
		for (int i = 0; i < n - 1; i++) {
			int minIdx = i;
			for (int j = i + 1; j < n; j++) {
				if (arr[j] < arr[minIdx]) minIdx = j;
			}
			int t = arr[i];
			arr[i] = arr[minIdx];
			arr[minIdx] = t;
		}
	}

	static void createInputFile() throws Exception {
		int n = readInt("Enter number of records (integers): ");
		if (n < 0) n = 0;
		BufferedWriter out = new BufferedWriter(new FileWriter(INPUT_FILE, false));
		System.out.println("Enter " + n + " integers:");
		for (int i = 0; i < n; i++) {
			int x = readInt("  value: ");
			out.write(Integer.toString(x));
			out.newLine();
		}
		out.close();
		System.out.println("Input file created: " + INPUT_FILE);
	}

	static void truncateFile(String path) throws Exception {
		new BufferedWriter(new FileWriter(path, false)).close();
	}

	static void distributeRuns(int m) throws Exception {
		File f = new File(INPUT_FILE);
		if (!f.exists()) {
			System.out.println("Input file not found. Create it first.");
			return;
		}
		BufferedReader in = new BufferedReader(new FileReader(f));
		BufferedWriter out1 = new BufferedWriter(new FileWriter(T1, false));
		BufferedWriter out2 = new BufferedWriter(new FileWriter(T2, false));
		int[] buf = new int[m];
		boolean toggle = true;

		while (true) {
			int k = 0;
			while (k < m) {
				Integer x = readNextInt(in);
				if (x == null) break;
				buf[k++] = x.intValue();
			}
			if (k == 0) break;
			selectionSort(buf, k);
			BufferedWriter out = toggle ? out1 : out2;
			for (int i = 0; i < k; i++) {
				out.write(Integer.toString(buf[i]));
				out.newLine();
			}
			toggle = !toggle;
		}
		in.close();
		out1.close();
		out2.close();
	}

	static boolean readNextInRun(BufferedReader in, int runSize, int[] countRead, int[] valueOut) throws Exception {
		if (countRead[0] >= runSize) return false;
		Integer x = readNextInt(in);
		if (x == null) return false;
		countRead[0]++;
		valueOut[0] = x.intValue();
		return true;
	}

	static void mergePass(String inAPath, String inBPath, String outAPath, String outBPath, int runSize) throws Exception {
		BufferedReader inA = new BufferedReader(new FileReader(inAPath));
		BufferedReader inB = new BufferedReader(new FileReader(inBPath));
		BufferedWriter outA = new BufferedWriter(new FileWriter(outAPath, false));
		BufferedWriter outB = new BufferedWriter(new FileWriter(outBPath, false));
		boolean toggle = true;

		while (true) {
			int[] countA = new int[] { 0 };
			int[] countB = new int[] { 0 };
			int[] aVal = new int[1];
			int[] bVal = new int[1];
			boolean hasA = readNextInRun(inA, runSize, countA, aVal);
			boolean hasB = readNextInRun(inB, runSize, countB, bVal);
			if (!hasA && !hasB) break;

			BufferedWriter out = toggle ? outA : outB;
			while (hasA || hasB) {
				if (!hasB || (hasA && aVal[0] <= bVal[0])) {
					out.write(Integer.toString(aVal[0]));
					out.newLine();
					hasA = readNextInRun(inA, runSize, countA, aVal);
				} else {
					out.write(Integer.toString(bVal[0]));
					out.newLine();
					hasB = readNextInRun(inB, runSize, countB, bVal);
				}
			}
			toggle = !toggle;
		}

		inA.close();
		inB.close();
		outA.close();
		outB.close();
	}

	static void copyFile(String src, String dst) throws Exception {
		BufferedReader in = new BufferedReader(new FileReader(src));
		BufferedWriter out = new BufferedWriter(new FileWriter(dst, false));
		Integer x;
		while ((x = readNextInt(in)) != null) {
			out.write(x.toString());
			out.newLine();
		}
		in.close();
		out.close();
	}

	static void externalSortMenu() throws Exception {
		int n = countRecords(INPUT_FILE);
		if (n <= 0) {
			System.out.println("Input file is empty or missing.");
			return;
		}
		int m = readInt("Enter m (records that fit in memory): ");
		if (m <= 0) {
			System.out.println("m must be positive.");
			return;
		}

		System.out.println("\nExternal Sort (2 input + 2 output tapes)");
		System.out.println("Records n = " + n + ", Memory m = " + m);

		distributeRuns(m);

		int runSize = m;
		int mergePasses = 0;
		String inA = T1, inB = T2, outA = T3, outB = T4;

		while (runSize < n) {
			mergePass(inA, inB, outA, outB, runSize);
			// swap tapes
			String tmpA = inA;
			inA = outA;
			outA = tmpA;
			String tmpB = inB;
			inB = outB;
			outB = tmpB;
			truncateFile(outA);
			truncateFile(outB);
			runSize *= 2;
			mergePasses++;
		}

		String finalTape = fileHasAnyInt(inA) ? inA : inB;
		copyFile(finalTape, OUTPUT_FILE);
		System.out.println("\nSorted output written to: " + OUTPUT_FILE);

		int totalPasses = 1 + mergePasses;
		System.out.println("Merge passes: " + mergePasses + ", Total passes (including initial): " + totalPasses);
		System.out.println("Time complexity (approx): O(n * log2(ceil(n/m))) merges, and O(n * passes) I/O.");
	}

	public static void main(String[] args) throws Exception {
		while (true) {
			System.out.println("\n===== Group F26 (External Sorting) =====");
			System.out.println("1. Create/overwrite input file");
			System.out.println("2. Display input file");
			System.out.println("3. Perform external sort");
			System.out.println("4. Display sorted output file");
			System.out.println("5. Exit");
			int ch = readInt("Enter choice: ");
			if (ch == 1) createInputFile();
			else if (ch == 2) displayFile(INPUT_FILE, "Input File");
			else if (ch == 3) externalSortMenu();
			else if (ch == 4) displayFile(OUTPUT_FILE, "Sorted Output");
			else if (ch == 5) break;
			else System.out.println("Invalid choice.");
		}
	}
}

File: DSAL/F/F26.cpp
// 26. Assume we have two input and two output tapes to perform the sorting. The internal memory can hold and sort m records at a time. Write a program in java for external sorting. Find out time complexity.

#include <fstream>
#include <iostream>
#include <string>
using namespace std;

static const char *INPUT_FILE = "F26_Input.txt";
static const char *T1 = "F26_T1.txt";
static const char *T2 = "F26_T2.txt";
static const char *T3 = "F26_T3.txt";
static const char *T4 = "F26_T4.txt";
static const char *OUTPUT_FILE = "F26_Sorted.txt";

void clearLine() {
  char c;
  while (cin.get(c)) {
    if (c == '\n') break;
  }
}

int readInt(const string &prompt) {
  int x;
  while (true) {
    cout << prompt;
    if (cin >> x) return x;
    cout << "Invalid number. Try again.\n";
    cin.clear();
    clearLine();
  }
}

bool fileHasAnyInt(const char *path) {
  ifstream in(path);
  int x;
  return static_cast<bool>(in >> x);
}

int countRecords(const char *path) {
  ifstream in(path);
  int x;
  int n = 0;
  while (in >> x) n++;
  return n;
}

void displayFile(const char *path, const string &title) {
  ifstream in(path);
  cout << "\n--- " << title << " ---\n";
  if (!in) {
    cout << "(file not found)\n";
    return;
  }
  int x;
  int shown = 0;
  while (in >> x) {
    cout << x << " ";
    shown++;
    if (shown % 20 == 0) cout << "\n";
  }
  if (shown == 0) cout << "(empty)";
  cout << "\nTotal records: " << shown << "\n";
}

void insertionSort(int arr[], int n) {
  for (int i = 1; i < n; ++i) {
    int key = arr[i];
    int j = i - 1;
    while (j >= 0 && arr[j] > key) {
      arr[j + 1] = arr[j];
      j--;
    }
    arr[j + 1] = key;
  }
}

void createInputFile() {
  int n = readInt("Enter number of records (integers): ");
  if (n < 0) n = 0;
  ofstream out(INPUT_FILE, ios::trunc);
  if (!out) {
    cout << "Unable to write input file.\n";
    return;
  }
  cout << "Enter " << n << " integers:\n";
  for (int i = 0; i < n; ++i) {
    int x = readInt("  value: ");
    out << x << "\n";
  }
  cout << "Input file created: " << INPUT_FILE << "\n";
}

void truncateFile(const char *path) {
  ofstream out(path, ios::trunc);
}

void distributeRuns(int m) {
  ifstream in(INPUT_FILE);
  if (!in) {
    cout << "Input file not found. Create it first.\n";
    return;
  }
  ofstream out1(T1, ios::trunc);
  ofstream out2(T2, ios::trunc);
  if (!out1 || !out2) {
    cout << "Unable to create tape files.\n";
    return;
  }

  int *buf = new int[m];
  bool toggle = true;
  while (true) {
    int k = 0;
    int x;
    while (k < m && (in >> x)) {
      buf[k++] = x;
    }
    if (k == 0) break;
    insertionSort(buf, k);
    ostream &out = toggle ? (ostream &)out1 : (ostream &)out2;
    for (int i = 0; i < k; ++i) out << buf[i] << "\n";
    toggle = !toggle;
    if (!in) break;
  }
  delete[] buf;
}

bool readNextInRun(ifstream &in, int runSize, int &countRead, int &value) {
  if (countRead >= runSize) return false;
  if (!(in >> value)) return false;
  countRead++;
  return true;
}

void mergePass(const char *inAPath, const char *inBPath, const char *outAPath, const char *outBPath, int runSize) {
  ifstream inA(inAPath);
  ifstream inB(inBPath);
  ofstream outA(outAPath, ios::trunc);
  ofstream outB(outBPath, ios::trunc);
  if (!outA || !outB) {
    cout << "Unable to open output tapes.\n";
    return;
  }
  bool toggle = true;

  while (true) {
    int countA = 0, countB = 0;
    int aVal = 0, bVal = 0;
    bool hasA = readNextInRun(inA, runSize, countA, aVal);
    bool hasB = readNextInRun(inB, runSize, countB, bVal);
    if (!hasA && !hasB) break;

    ostream &out = toggle ? (ostream &)outA : (ostream &)outB;
    while (hasA || hasB) {
      if (!hasB || (hasA && aVal <= bVal)) {
        out << aVal << "\n";
        hasA = readNextInRun(inA, runSize, countA, aVal);
      } else {
        out << bVal << "\n";
        hasB = readNextInRun(inB, runSize, countB, bVal);
      }
    }
    toggle = !toggle;
  }
}

void copyFile(const char *src, const char *dst) {
  ifstream in(src);
  ofstream out(dst, ios::trunc);
  int x;
  while (in >> x) out << x << "\n";
}

void externalSortMenu() {
  int n = countRecords(INPUT_FILE);
  if (n <= 0) {
    cout << "Input file is empty or missing.\n";
    return;
  }
  int m = readInt("Enter m (records that fit in memory): ");
  if (m <= 0) {
    cout << "m must be positive.\n";
    return;
  }

  cout << "\nExternal Sort (2 input + 2 output tapes)\n";
  cout << "Records n = " << n << ", Memory m = " << m << "\n";

  // 1) Initial distribution of sorted runs (size <= m) onto two input tapes.
  distributeRuns(m);

  int runSize = m;
  int mergePasses = 0;
  const char *inA = T1;
  const char *inB = T2;
  const char *outA = T3;
  const char *outB = T4;

  while (runSize < n) {
    mergePass(inA, inB, outA, outB, runSize);
    // swap tapes
    const char *tmpA = inA;
    inA = outA;
    outA = tmpA;
    const char *tmpB = inB;
    inB = outB;
    outB = tmpB;
    // clear old output (now used as output in next pass)
    truncateFile(outA);
    truncateFile(outB);
    runSize *= 2;
    mergePasses++;
  }

  const char *finalTape = fileHasAnyInt(inA) ? inA : inB;
  copyFile(finalTape, OUTPUT_FILE);
  cout << "\nSorted output written to: " << OUTPUT_FILE << "\n";

  int totalPasses = 1 + mergePasses; // initial distribution + merge passes
  cout << "Merge passes: " << mergePasses << ", Total passes (including initial): " << totalPasses << "\n";
  cout << "Time complexity (approx): O(n * log2(ceil(n/m))) merges, and O(n * passes) I/O.\n";
}

int main() {
  while (true) {
    cout << "\n===== Group F26 (External Sorting) =====\n";
    cout << "1. Create/overwrite input file\n";
    cout << "2. Display input file\n";
    cout << "3. Perform external sort\n";
    cout << "4. Display sorted output file\n";
    cout << "5. Exit\n";
    int ch = readInt("Enter choice: ");
    if (ch == 1) createInputFile();
    else if (ch == 2) displayFile(INPUT_FILE, "Input File");
    else if (ch == 3) externalSortMenu();
    else if (ch == 4) displayFile(OUTPUT_FILE, "Sorted Output");
    else if (ch == 5) break;
    else cout << "Invalid choice.\n";
  }
  return 0;
}
